Sources/SwiftProtobuf/TextScanner.swift

```diff
@@ -15,55 +15,227 @@
+        case 98: // \b
+          out.append(8)
+        case 102: // \f
+          out.append(12)
+        case 110: // \n
+          out.append(10)
+        case 114: // \r
+          out.append(13)
+        case 116: // \t
+          out.append(9)
+        case 118: // \v
+          out.append(11)
+        case 34, 39, 63, 92: // " ' ? \
+          out.append(escaped)
+        default:
+          return nil // Unrecognized escape
+        }
+      } else {
+        return nil // Input ends with backslash
+      }
+    default:
+      out.append(byte)
+    }
+  }
+  // There has got to be an easier way to convert a [UInt8] into a String.
+  out.append(0)
+  return out.withUnsafeBufferPointer { ptr in
+    if let addr = ptr.baseAddress {
+      return addr.withMemoryRebound(to: CChar.self, capacity: ptr.count) { p in
+        let q = UnsafePointer<CChar>(p)
+        let s = String(validatingUTF8: q)
+        return s
+      }
+    } else {
+      return ""
+    }
+  }
+}
+
 ///
 /// TextScanner has no public members.
 ///
 public class TextScanner {
     internal var extensions: ExtensionSet?
-    private var scalars: String.UnicodeScalarView
-    private var index: String.UnicodeScalarView.Index
-    private var tokenStart: String.UnicodeScalarView.Index
-    private var tokenPushback = [TextToken]()
+    private var utf8: String.UTF8View
+    private var index: String.UTF8View.Index
     private var eof: Bool = false
     internal var complete: Bool {
-        while index != scalars.endIndex {
-            let c = scalars[index]
-            switch c {
-            case " ", "\t", "\r", "\n":
-                index = scalars.index(after: index)
-            default:
-                return false
-            }
-        }
-        return true
+        skipWhitespace()
+        return index == utf8.endIndex
     }
 
     internal init(text: String, extensions: ExtensionSet? = nil) {
-        scalars = text.unicodeScalars
-        index = scalars.startIndex
-        tokenStart = index
+        utf8 = text.utf8
+        index = utf8.startIndex
         self.extensions = extensions
     }
 
-    internal func pushback(token: TextToken) {
-        tokenPushback.append(token)
-    }
-
-    /// Skip whitespace, set token start to first non-whitespace character
+    /// Skip whitespace
     private func skipWhitespace() {
         var lastIndex = index
-        while index != scalars.endIndex {
-            let scalar = scalars[index]
-            switch scalar {
-            case " ", "\t", "\r", "\n":
-                index = scalars.index(after: index)
+        while index != utf8.endIndex {
+            let u = utf8[index]
+            switch u {
+            case 32, 9, 10, 13: // space, tab, NL, CR
+                index = utf8.index(after: index)
                 lastIndex = index
-            case "#":
-                while index != scalars.endIndex {
+            case 35: // #
+                while index != utf8.endIndex {
```

Comment:
Do you want to advance once before the loop?  Otherwise, you check fetch/check the '#' again, no?